### PR TITLE
Design System: Add Unit + Integer variants to input

### LINKED ISSUE #6414: Design System: Input Variants - Number & Unit
## Feature Description

Add remaining input variants. 
[Designs](https://www.figma.com/file/bMhG3KyrJF8vIAODgmbeqT/Design-System?node-id=3397%3A800)

### Unit
Sometimes inputs will have a unit suffixed (or technically prefixed if that's a thing in some languages) but it's removed on data entry (focus). This happens for `%` for opacity, `°` (degree) for rotation, and `ms` for durations (in the animation panel).

The current implementation calls this `symbol`, but I tend to think that `unit` (default `''`) might be a better name

### Number
Numbers are a bit special as we do several magic things today to make them easier to work with.

* Only parse the input to a number on blur, enter or arrow key (allow invalid input until then) and only then submit up the chain
* ⬆️   and ⬇️   changes the value 1 up or down
* Some inputs will allow negative numbers, others won't

We have this as a completely separate input field type (`<Numeric />` rather than `<Text />`) with the property `canBeNegative` (default `false`) today, but I think something like `isNumeric` and `canBeNegative` (both default `false`) works better.

### Allow empty
Some inputs revert to their previous value if emptied and blurred, while others are allowed to be empty on blur and thus submit `null` up the chain.

We currently have this as `canBeEmpty` (default `false`).

### Float or integer
We actually also have a boolean flag for telling whether only integers or also floats are allowed. Currently this is called `float`, but `isFloat` (default `false`) is probably better aligned with existing props.

## Alternatives Considered

n/a

## Additional Context

The editor already has this built in to the input it is using. We need to add these variants to the input in the design system: https://github.com/google/web-stories-wp/pull/6355#issuecomment-780749936

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by samwhale**
## Context

Building out the design system. These updates for the input are necessary in the editor.

## Summary

Adds the `Unit` and `Numeric` input variants as described in #6414 and this comment: https://github.com/google/web-stories-wp/pull/6355#issuecomment-780749936

## Relevant Technical Choices

The `Input` component is meant to be a 'presentational' component. However, the functionality of 'reverting to a previous value' makes this component more of a 'container' component.

I am keeping the `Input` as presentational as possible, but wrapping it in a container component to add functionality that the 'numeric' input will need (saving old inputs, only calling `onChange` when the value should be "submitted", etc.). Other changes:
- `allowNegative` is redundant when paired with `max` and `min`. Removed it so that we use only max/min. If there is a bound, we can specify it there.
- I kept the API similar to the old input rather than using `type="numeric"` for the input. I would prefer to use `type="numeric"` but there was already old logic written that I used instead of making a brand new component
- created this without setting event listeners on the window. This should be more performant and relies on only "react"
- added a hook to manage the `Numeric` input's state. This also makes testing easier.

## To-do

After this PR is merged I'd like to make one final PR that merges `Input` and `Numeric` into one component. That component will decide which input type to use based on a `type` prop.
- if `type="numeric"` - render the `<Numeric />` input
- else render the base `<Input />`

## User-facing changes

none

## Testing Instructions

Visit storybook. Design System -> Input -> Numeric

Test the following scenarios:
- can enter any value while typing
- blurring the field will trigger validation. If there is a valid number, the number will remain.
    - if the value is invalid and the previous value was valid, leave the valid number
    - if the value is invalid and the previous value was also invalid, leave a new valid number
- `isFloat=true` will allow for float values as well as integers
- arrow keys: ⬆️  and ⬇️  changes the value 1 up or down
- arrow keys + alt: ⬆️  and ⬇️  changes the value by 0.1 up or down (if `isFloat` is `true`)
- `unit` should only be visible when the field is not focused
- field cannot be empty unless `allowEmpty` is `true`

All of these values can be configured in the `knobs` section in storybook

### QA

- [x] This is a non-user-facing change and requires no QA

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

Visit storybook. Design System -> Input -> Numeric

Test the following scenarios:
- can enter any value while typing
- blurring the field will trigger validation. If there is a valid number, the number will remain.
    - if the value is invalid and the previous value was valid, leave the valid number
    - if the value is invalid and the previous value was also invalid, leave a new valid number
- `isFloat=true` will allow for float values as well as integers
- arrow keys: ⬆️  and ⬇️  changes the value 1 up or down
- arrow keys + alt: ⬆️  and ⬇️  changes the value by 0.1 up or down (if `isFloat` is `true`)
- `unit` should only be visible when the field is not focused
- field cannot be empty unless `allowEmpty` is `true`

All of these values can be configured in the `knobs` section in storybook

|||
|--|--|
|Numeric input|<image src="https://user-images.githubusercontent.com/22185279/108396951-aaf32900-71d4-11eb-9c39-6e30de3dee11.gif" width="400px" />|
|Unit|<image src="https://user-images.githubusercontent.com/22185279/108396946-a9296580-71d4-11eb-8528-a2480895e020.gif" width="400px" />|
|Minimum|<image src="https://user-images.githubusercontent.com/22185279/108396933-a62e7500-71d4-11eb-8850-61442f15dd20.gif" width="200px" />|
|Maximum|<image src="https://user-images.githubusercontent.com/22185279/108396939-a890cf00-71d4-11eb-8d02-fc4362078cbf.gif" width="200px" />|
|`ArrowUp` and `ArrowDown`|<image src="https://user-images.githubusercontent.com/22185279/108396947-a9c1fc00-71d4-11eb-8ba7-62d40420f3eb.gif" width="400px" />|
|`Esc` key press|<image src="https://user-images.githubusercontent.com/22185279/108396949-aa5a9280-71d4-11eb-9a11-d9e2ac5df671.gif" width="400px" />|

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6414

### CODE REVIEW

**Comment by samwhale** [comment]
Note: when testing this and changing the `max` and `min` knobs in storybook you will see this error:
image

When testing this, `value` was set the whole time and didn't get removed. I don't think this component is uncontrolled at all, but if you see something strange lmk.

I think the above error is a storybook 🐛

File: assets/src/design-system/components/input/test/useNumericInput.js [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+      ${1.01}      | ${'ArrowUp'}   | ${true}  | ${true}  | ${undefined} | ${undefined} | ${1.11}
+      ${1.01}      | ${'ArrowDown'} | ${true}  | ${true}  | ${undefined} | ${undefined} | ${0.91}
+      ${10}        | ${'ArrowUp'}   | ${false} | ${false} | ${10}        | ${undefined} | ${10}
+      ${-2}        | ${'ArrowDown'} | ${false} | ${false} | ${undefined} | ${-2}        | ${-2}
+      ${10}        | ${'ArrowUp'}   | ${false} | ${true}  | ${10}        | ${undefined} | ${10}
+      ${-2}        | ${'ArrowDown'} | ${false} | ${true}  | ${undefined} | ${-2}        | ${-2}
+      ${10}        | ${'ArrowUp'}   | ${true}  | ${false} | ${10}        | ${undefined} | ${10}
+      ${-2}        | ${'ArrowDown'} | ${true}  | ${false} | ${undefined} | ${-2}        | ${-2}
+      ${10}        | ${'ArrowUp'}   | ${true}  | ${true}  | ${10}        | ${undefined} | ${10}
+      ${-2}        | ${'ArrowDown'} | ${true}  | ${true}  | ${undefined} | ${-2}        | ${-2}
```
Comment (barklund) [inline assets/src/design-system/components/input/test/useNumericInput.js:175]: Seriously thorough effort here - 100 points to Gryffindor for this work! 🎉 🎉 🎉

File: assets/src/design-system/components/input/index.js [hunk 2]
```diff
@@ -117,60 +124,90 @@ const StyledInput = styled.input(
...
+    },
+    ref
+  ) => {
+    const inputId = useMemo(() => id || uuidv4(), [id]);
+    const inputRef = useRef(null);
+    const [focused, setFocused] = useState(false);
+
+    useEffect(() => {
+      // focus input when focused state is set
+      if (focused && ref && ref.current) {
```
Comment (BrittanyIRL) [inline assets/src/design-system/components/input/index.js]: nit: could do `ref?.current` to combine `ref && ref.current`

File: assets/src/design-system/components/input/index.js [hunk 3]
```diff
@@ -117,60 +124,90 @@ const StyledInput = styled.input(
...
+        ref.current.select();
+      } else if (focused && inputRef.current) {
+        inputRef.current.select();
+      }
```
Comment (BrittanyIRL) [inline assets/src/design-system/components/input/index.js]: nit: could do 
```
else if (focused) {
inputRef?.current.select();
}
```

File: assets/src/design-system/components/input/useNumericInput.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        newValue = parseInput(currentValue, options);
+      }
+
+      revertToOriginal.current = false;
+
+      setCurrentValue(newValue);
+      onChange(ev, newValue);
+    },
+    [currentValue, onChange, options]
+  );
```
Comment (barklund) [inline assets/src/design-system/components/input/useNumericInput.js]: The only issue I can find is here - if you type something illegal and blur or press enter (both are routed through this callback), the input is left as "null", where it instead should revert to the original text ("last known valid state"). Otherwise, you can get his confusing output of `nullms`:

invalid-input-cut

**Review by barklund — APPROVED** [review]
Impressive effort here with excellent SoC and DRY. And thanks for the quick turn-around too - this will come in handy next week!

File: assets/src/design-system/components/input/test/useNumericInput.js [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        result.current.handleBlur('dummy data');
+      });
+      expect(mockOnChange).toHaveBeenCalledTimes(1);
+      expect(mockOnChange).toHaveBeenCalledWith('dummy data', 20);
+    });
+  });
+
+  describe('handleKeyUpDown', () => {
+    it.each`
+      initialValue | key            | altKey   | isFloat  | max          | min          | finalValue
```
Comment (BrittanyIRL) [inline assets/src/design-system/components/input/test/useNumericInput.js:159]: This is real nice

**Review by BrittanyIRL — CHANGES_REQUESTED** [review]
This is really really well done! Keyboard works great. I love the unit tests and the code is really easy to read. 2 edge cases of just bad data getting entered but we should probably watch for? 

1. If I enter some alpa characters and blur, it returns to previously set 0. When i come back, clear the 0 and type some more alpha it turns to `null` on blur. 
null

2. If I enter in `.7` on `unit and suffix` i get `nullms` on blur 
decimal null

**Comment by samwhale** [comment]
> The only issue I can find is here - if you type something illegal and blur or press enter (both are routed through this callback), the input is left as "null", where it instead should revert to the original text ("last known valid state"). Otherwise, you can get his confusing output of `nullms`:

Great catch @BrittanyIRL and @barklund. Thanks for such a thorough review, and for reviewing such a large pr. I greatly appreciate it.

I fixed the bug! It now reverts to the original value if the new value is invalid. This commit fixed it: https://github.com/google/web-stories-wp/pull/6419/commits/5a2d108f0e436caf98009d17be1f56dccb73fe8c. However, it assumes that the original value was valid, which I think would be a fair assumption.

If we want to really guard against bad input values, we should probably add a default/fallback value 🤔 

buggy-boi-fix

File: assets/src/design-system/components/input/useNumericInput.js (hunk 4 above)
Comment (samwhale) [reply assets/src/design-system/components/input/useNumericInput.js]: Sorry about the force push - I rebased `main`.

The new commit is this: https://github.com/google/web-stories-wp/pull/6419/commits/5a2d108f0e436caf98009d17be1f56dccb73fe8c

**Review by BrittanyIRL — APPROVED** [review]
Awesome! 🚢

**Review by barklund — APPROVED** [review]
